server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/simplecoin/sources/SimpleCoinRateSource.java

```diff
@@ -0,0 +1,149 @@
+    private ISimpleCoinApi api;
+    private static HashMap<String, BigDecimal> rateAmounts = new HashMap<>();
+    private static HashMap<String, Long> rateTimes = new HashMap<>();
+    private static final long MAXIMUM_ALLOWED_TIME_OFFSET = 30 * 1000; //30sec
+    private SupportedCurrencies supportedCurrencies;
+
+    public SimpleCoinRateSource(SupportedCurrencies currencies) {
+        this.supportedCurrencies = currencies;
+        api = RestProxyFactory.createProxy(ISimpleCoinApi.class, "https://server.simplecoin.eu");
+    }
+
+    @Override
+    public BigDecimal getExchangeRateForBuy(String cryptoCurrency, String fiatCurrency) {
+        return calculateBuyPrice(cryptoCurrency, fiatCurrency, BigDecimal.ONE);
+    }
+
+    @Override
+    public BigDecimal calculateBuyPrice(String cryptoCurrency, String fiatCurrency, BigDecimal cryptoAmount) {
+        try {
+            RateLimiter.waitForPossibleCall(getClass());
+            FiatCryptoResponse ticker = api.returnRate(fiatCurrency, cryptoCurrency); // sequence for buying crypto (e.g. from=CZK to=BTC)
+            BigDecimal price = ticker.getResponse().getRate();// for buying attribute rate
+            log.warn("Called buy rate: {}{} = {}", cryptoCurrency, fiatCurrency, price);
+            return price;
+
+        } catch (Throwable e) {
+            log.warn("Error", e);
+        }
+        return null;
+    }
+
+    @Override
+    public BigDecimal getExchangeRateForSell(String cryptoCurrency, String fiatCurrency) {
+        return calculateSellPrice(cryptoCurrency, fiatCurrency, BigDecimal.ONE);
+    }
+
+    @Override
+    public BigDecimal calculateSellPrice(String cryptoCurrency, String fiatCurrency, BigDecimal cryptoAmount) {
+        try {
+            RateLimiter.waitForPossibleCall(getClass());
+            FiatCryptoResponse ticker = api.returnRate(cryptoCurrency, fiatCurrency); // sequence for selling crypto (e.g. from=BTC to=CZK)
+            BigDecimal price = ticker.getResponse().getRate_inverse(); // for selling rate_inverse
+            log.warn("Called sell rate: {}{} = {}", cryptoCurrency, fiatCurrency, price);
+            return price;
+
+        } catch (Throwable e) {
+            log.warn("Error", e);
+        }
+        return null;
+    }
+
+    @Override
+    public Set<String> getFiatCurrencies() {
+        return supportedCurrencies.getSupportedFiatCurrency();
+    }
+
+    @Override
+    public String getPreferredFiatCurrency() {
+        return supportedCurrencies.getPreferredFiatCurrency();
+    }
+
+    @Override
+    public Set<String> getCryptoCurrencies() {
+        return supportedCurrencies.getSupportedCryptoCurrency();
+    }
+
+    private boolean isCurrencySupported(String cryptoCurrency, String fiatCurrency) {
+        if (supportedCurrencies.isCryptoSupported(cryptoCurrency)
+            && supportedCurrencies.isFiatSupported(fiatCurrency)) {
+            return true;
+        }
+        log.debug("Unsupported currency");
```

Comment:
Useless log message - what is unsupported? Add `cyptoCurrency` and `fiatCurrency` to log message or remove this line